plugins/module_utils/identity/keycloak/keycloak.py

```diff
@@ -280,6 +311,46 @@ def __init__(self, module, connection_header):
         self.restheaders = connection_header
         self.http_agent = self.module.params.get('http_agent')
 
+    def _request(self, url, method, data=None):
+        def make_request_catching_401():
+            try:
+                return open_url(url, method=method, data=data,
+                                http_agent=self.http_agent, headers=self.restheaders,
+                                timeout=self.connection_timeout,
+                                validate_certs=self.validate_certs)
+            except HTTPError as e:
+                if e.code != 401:
+                    raise e
+                return e
+
+        r = make_request_catching_401()
+        if not isinstance(r, Exception):
+            return r
+
```

Comment:
Little late in the game for that, I know, but I believe this would be leaner as:
```suggestion
    def _request(self, url, method, data=None):
        class Unauthorized(Exception):
            pass

        def make_request_catching_401():
            try:
                return open_url(url, method=method, data=data,
                                http_agent=self.http_agent, headers=self.restheaders,
                                timeout=self.connection_timeout,
                                validate_certs=self.validate_certs)
            except HTTPError as e:
                if e.code != 401:
                    raise e
                raise Unauthorized()

        try:
            return make_request_catching_401()
        exception Unauthorized:
            pass
        except Exception:
            raise
```
and so on

Reply:
I like this, however we need to keep the original 401 exception in case no re-auth options are available and it needs to be re-raised. (Thanks for catching the mistake where it was being returned and not raised in your other comment.)